src/quota_aggregator_impl.cc

```diff
@@ -0,0 +1,248 @@
+==============================================================================*/
+
+#include <iostream>
+
+#include "src/quota_aggregator_impl.h"
+#include "src/signature.h"
+
+#include "google/protobuf/stubs/logging.h"
+#include "google/protobuf/text_format.h"
+
+using std::string;
+using ::google::api::MetricDescriptor;
+using ::google::api::servicecontrol::v1::QuotaOperation;
+using ::google::api::servicecontrol::v1::AllocateQuotaRequest;
+using ::google::api::servicecontrol::v1::AllocateQuotaResponse;
+using ::google::protobuf::util::Status;
+using ::google::protobuf::util::error::Code;
+using ::google::service_control_client::SimpleCycleTimer;
+
+namespace google {
+namespace service_control_client {
+
+void QuotaAggregatorImpl::CacheElem::Aggregate(
+    const AllocateQuotaRequest& request) {
+  if (operation_aggregator_ == NULL) {
+    operation_aggregator_.reset(
+        new QuotaOperationAggregator(request.allocate_operation()));
+  } else {
+    operation_aggregator_->MergeOperation(request.allocate_operation());
+  }
+}
+
+AllocateQuotaRequest
+QuotaAggregatorImpl::CacheElem::ReturnAllocateQuotaRequestAndClear(
+    const string& service_name, const std::string& service_config_id) {
+  AllocateQuotaRequest request;
+
+  request.set_service_name(service_name);
+  request.set_service_config_id(service_config_id);
+
+  if (operation_aggregator_ != NULL) {
+    *(request.mutable_allocate_operation()) =
+        operation_aggregator_->ToOperationProto();
+    operation_aggregator_ = NULL;
+  }
+
+  return request;
+}
+
+QuotaAggregatorImpl::QuotaAggregatorImpl(const std::string& service_name,
+                                         const std::string& service_config_id,
+                                         const QuotaAggregationOptions& options)
+    : service_name_(service_name),
+      service_config_id_(service_config_id),
+      options_(options) {
+  if (options.num_entries > 0) {
+    cache_.reset(new QuotaCache(
+        options.num_entries, std::bind(&QuotaAggregatorImpl::OnCacheEntryDelete,
+                                       this, std::placeholders::_1)));
+    cache_->SetAgeBasedEviction(options.refresh_interval_ms / 1000.0);
+  }
+}
+
+QuotaAggregatorImpl::~QuotaAggregatorImpl() {
+  SetFlushCallback(NULL);
+  FlushAll();
+}
+
+// Sets the flush callback function.
+// The callback function must be light and fast.  If it needs to make
+// a remote call, it must be non-blocking call.
+// It should NOT call into this object again from this callback.
+// It will cause dead-lock.
+void QuotaAggregatorImpl::SetFlushCallback(FlushCallback callback) {
+  InternalSetFlushCallback(callback);
+}
+
+// If the quota could not be handled by the cache, returns NOT_FOUND,
+// caller has to send the request to service control.
+// Otherwise, returns OK and cached response.
+::google::protobuf::util::Status QuotaAggregatorImpl::Quota(
+    const ::google::api::servicecontrol::v1::AllocateQuotaRequest& request,
```

Comment:
Could we add comment to describe our cache algorithm?  

It will be easier for others to read the code.